File: services/node/src/services/order.ts

import type { FastifyInstance } from "fastify";
import type { Pool } from "pg";
import type { Config } from "../config";

interface OrderRow {
  id: string;
  user_id: string;
  product_id: string;
  qty: number;
}

interface Order {
  id: number;
  user_id: number;
  product_id: number;
  qty: number;
}

interface CreateBody {
  user_id?: number | string;
  product_id?: number | string;
  qty?: number | string;
}

function toOrder(row: OrderRow): Order {
  return {
    id: Number(row.id),
    user_id: Number(row.user_id),
    product_id: Number(row.product_id),
    qty: row.qty,
  };
}

async function exists(url: string, timeoutMs: number): Promise<boolean> {
  try {
    const res = await fetch(url, {
      method: "GET",
      signal: AbortSignal.timeout(timeoutMs),
    });
    await res.arrayBuffer().catch(() => undefined);
    return res.ok;
  } catch {
    return false;
  }
}

export function registerOrderRoutes(
  app: FastifyInstance,
  pool: Pool,
  cfg: Config,
): void {
  app.post<{ Body: CreateBody }>("/orders", async (request, reply) => {
    const { user_id, product_id, qty } = request.body ?? {};
    if (user_id === undefined || product_id === undefined || qty === undefined) {
      return reply
        .code(400)
        .send({ error: "user_id, product_id and qty are required" });
    }

    const userUrl = `${cfg.userServiceUrl}/users/${user_id}`;
    const productUrl = `${cfg.productServiceUrl}/products/${product_id}`;

    const [userOk, productOk] = await Promise.all([
      exists(userUrl, cfg.fanoutTimeoutMs),
      exists(productUrl, cfg.fanoutTimeoutMs),
    ]);

    if (!userOk) {
      return reply.code(400).send({ error: "user does not exist" });
    }
    if (!productOk) {
      return reply.code(400).send({ error: "product does not exist" });
    }

    const { rows } = await pool.query<OrderRow>(
      `INSERT INTO orders (user_id, product_id, qty)
       VALUES ($1, $2, $3)
       RETURNING id, user_id, product_id, qty`,
      [user_id, product_id, qty],
    );
    return reply.code(201).send(toOrder(rows[0]));
  });

  app.get<{ Params: { id: string } }>("/orders/:id", async (request, reply) => {
    const { rows } = await pool.query<OrderRow>(
      "SELECT id, user_id, product_id, qty FROM orders WHERE id = $1",
      [request.params.id],
    );
    const row = rows[0];
    if (!row) {
      return reply.code(404).send({ error: "not found" });
    }
    return reply.send(toOrder(row));
  });
}

File: services/go/internal/order/order.go
package order

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
)

type Service struct {
	db                *sql.DB
	client            *http.Client
	userServiceURL    string
	productServiceURL string
}

func New(db *sql.DB, userServiceURL, productServiceURL string) *Service {
	return &Service{
		db:                db,
		client:            &http.Client{Timeout: 5 * time.Second},
		userServiceURL:    userServiceURL,
		productServiceURL: productServiceURL,
	}
}

func (s *Service) Register(r gin.IRouter) {
	r.POST("/orders", s.create)
	r.GET("/orders/:id", s.getByID)
}

type createBody struct {
	UserID    int64 `json:"user_id"`
	ProductID int64 `json:"product_id"`
	Qty       int   `json:"qty"`
}

func (s *Service) create(c *gin.Context) {
	var body createBody
	if err := c.ShouldBindJSON(&body); err != nil || body.UserID == 0 || body.ProductID == 0 || body.Qty <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "user_id, product_id and positive qty are required"})
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), 5*time.Second)
	defer cancel()

	userURL := fmt.Sprintf("%s/users/%d", s.userServiceURL, body.UserID)
	productURL := fmt.Sprintf("%s/products/%d", s.productServiceURL, body.ProductID)

	errCh := make(chan error, 2)
	go func() { errCh <- s.checkExists(ctx, userURL, "user") }()
	go func() { errCh <- s.checkExists(ctx, productURL, "product") }()

	var fanoutErr error
	for i := 0; i < 2; i++ {
		if err := <-errCh; err != nil && fanoutErr == nil {
			fanoutErr = err
		}
	}
	if fanoutErr != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": fanoutErr.Error()})
		return
	}

	var id int64
	err := s.db.QueryRowContext(c.Request.Context(),
		`INSERT INTO orders (user_id, product_id, qty) VALUES ($1, $2, $3) RETURNING id`,
		body.UserID, body.ProductID, body.Qty,
	).Scan(&id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "insert failed"})
		return
	}

	c.JSON(http.StatusCreated, gin.H{
		"id":         id,
		"user_id":    body.UserID,
		"product_id": body.ProductID,
		"qty":        body.Qty,
	})
}

func (s *Service) checkExists(ctx context.Context, url, kind string) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return fmt.Errorf("%s lookup failed", kind)
	}
	resp, err := s.client.Do(req)
	if err != nil {
		return fmt.Errorf("%s service unreachable", kind)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		return nil
	}
	return fmt.Errorf("%s not found", kind)
}

func (s *Service) getByID(c *gin.Context) {
	id := c.Param("id")
	var (
		oid       int64
		userID    int64
		productID int64
		qty       int
	)
	err := s.db.QueryRowContext(c.Request.Context(),
		`SELECT id, user_id, product_id, qty FROM orders WHERE id = $1`, id,
	).Scan(&oid, &userID, &productID, &qty)
	if errors.Is(err, sql.ErrNoRows) {
		c.JSON(http.StatusNotFound, gin.H{"error": "not found"})
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "query failed"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"id":         oid,
		"user_id":    userID,
		"product_id": productID,
		"qty":        qty,
	})
}
